site/app/controllers/admin/PlagiarismController.php

```diff
@@ -533,161 +636,280 @@ public function saveNewPlagiarismConfiguration(string $new_or_edit, string $grad
+                    $config_id = 1;
+                }
+                else {
+                    $config_id = $config_id->getConfigID() + 1;
+                }
+            }
+            catch (Exception $e) {
+                $this->core->addErrorMessage($e->getMessage());
                 return new RedirectResponse($return_url);
             }
         }
 
+        // Save configuration form data
+        if ($new_or_edit === "new" && isset($_POST["config_option"]) && $_POST["config_option"] === "upload_config") { // uploaded config file
+            // error checking
+            if (empty($_FILES) || !isset($_FILES["upload_config_file"]) || !isset($_FILES["upload_config_file"]["tmp_name"]) || $_FILES["upload_config_file"]["tmp_name"] === "") {
+                $this->core->addErrorMessage("Error: File upload failed");
+                return new RedirectResponse($return_url);
+            }
+            // load, parse, and save the config info
+            try {
+                $data = json_decode(file_get_contents($_FILES["upload_config_file"]["tmp_name"]), true);
 
-        // Version /////////////////////////////////////////////////////////////
-        $version_option = $_POST['version_option'];
-        assert($version_option == "active_version" || $version_option == "all_versions");
+                // This is a little ugly/repetitive but it can be frustrating for users to get nondescriptive errors
+                // so we try to make potential error cases a little more helpful
+                $error_message = "";
+                if (!isset($data["version"])) {
+                    $error_message = "Error: Invalid or missing version field";
+                }
+                elseif (!isset($data["regex"])) {
+                    $error_message = "Error: Invalid or missing regex field";
+                }
+                elseif (!isset($data["regex_dirs"])) {
+                    $error_message = "Error: Invalid or missing regex_dirs field";
+                }
+                elseif (!isset($data["language"])) {
+                    $error_message = "Error: Invalid or missing language field";
+                }
+                elseif (!isset($data["threshold"])) {
+                    $error_message = "Error: Invalid or missing threshold field";
+                }
+                elseif (!isset($data["sequence_length"])) {
+                    $error_message = "Error: Invalid or missing sequence_length field";
+                }
+                elseif (!isset($data["prior_term_gradeables"])) {
+                    $error_message = "Error: Invalid or missing prior_term_gradeables field";
+                }
+                elseif (!isset($data["ignore_submissions"])) {
+                    $error_message = "Error: Invalid or missing ignore_submissions field";
+                }
 
+                if ($error_message !== "") {
+                    throw new Exception($error_message);
```

Comment:
What other exceptions might we hit within this try/catch? The validation errors seems like something a human might accidentally run into, but other exceptions (e.g. coming from the DB) should probably not be caught so that we (the developers) get an error report of it happening as it shouldn't ever happen.

Reply:
Hmm, tricky.  Perhaps there is a better way to do it but I moved a decent bit of validation to `PlagiarismConfig.php` so that as much input validation as possible is done in that one central place.  The main thing I was trying to catch here were potential validation errors (such as when an invalid `config.json` is uploaded) which should be referred back to the user.  Thoughts?

Reply:
Use more specific exception classes, in this case [`ValidationException`](https://github.com/Submitty/Submitty/blob/master/site/app/exceptions/ValidationException.php) seems appropriate. Then where the catches are you would catch just these types of exceptions.

Generally, in PHP (and other languages with typed exceptions), you want to avoid throwing a broad exception or catching a broad exception, preferring to make things as scoped as possible. Python is a better example of why to use narrow exceptions, where any error including syntax errors are thrown exceptions.